arch/x86_64/mcount-insn.c

```diff
@@ -78,6 +78,79 @@ static int opnd_reg(int capstone_reg)
 	return -1;
 }
 
+/*
+ * Handle relative jumps and conditional branches.
+ *
+ * This function relocates jcc8 and jcc32 instructions by replacing them with a jcc8 
+ * that has a null offset. The offset will be patched later when the code is saved for out
+ * of line execution. The new jcc8 will jump (if condition is met) to a trampoline
+ * which itself jumps to the target of the orginal instruction.
+ *
+ * The relocation of jmp8 and jmp32 is achieved by replacing them with an absolute 
+ * indirect jump to the target.
+ *
+ */
+static int handle_rel_jmp(cs_insn *insn, uint8_t insns[], struct mcount_disasm_info *info)
+{
+	cs_x86 *x86 = &insn->detail->x86;
+	uint8_t relocated_insn[15] = { 0xff, 0x25,};
+	uint8_t opcode = insn->bytes[0];
+	uint64_t target;
+	struct cond_branch_info *cbi;
+
+#define JMP8_OPCODE 0xEB
+#define JMP32_OPCODE 0xE9
+#define JCC8_SIZE   2
+#define IND_JMP_SIZE 6
+#define OP   0
+#define OFS	  1
+	cs_x86_op *opnd = &x86->operands[0];
+
+	if (x86->op_count != 1 || opnd->type != X86_OP_IMM)
+		goto out;
+
+	if ((opcode == JMP8_OPCODE || opcode == JMP32_OPCODE)) {
+		if (strcmp(insn->mnemonic, "jmp") != 0)
+			goto out;
+
+		target = opnd->imm;
+
+		memcpy(insns, relocated_insn, IND_JMP_SIZE);
+		memcpy(insns + IND_JMP_SIZE, &target, sizeof(target));
+		return IND_JMP_SIZE + sizeof(target);
+
+	}/* Jump relative 8 if condition is met (except JCXZ, JECXZ and JRCXZ) */
+	else if((opcode & 0xF0) == 0x70) {
+
+		cbi = &info->branch_info[info->nr_branch++];
+		cbi->insn_index = info->copy_size;
+		cbi->branch_target = opnd->imm;
+
+		relocated_insn[OP] = opcode;
+		relocated_insn[OFS] = 0x00;
+
+		memcpy(insns, (void *)relocated_insn, JCC8_SIZE);
+		return JCC8_SIZE;
+
+	}/* Jump relative 32 if condition is met */
+	else if(opcode == 0x0F && (insn->bytes[1] & 0xF0) == 0x80) {
+
+		cbi = &info->branch_info[info->nr_branch++];
+		cbi->insn_index = info->copy_size;
+		cbi->branch_target = opnd->imm;
+
+		/* We use the equivalent jcc8 of the original jcc32 */
+		relocated_insn[OP] = insn->bytes[1] - 0x10;
+		relocated_insn[OFS] = 0x00;
```

Comment:
> Hmm.. can we keep jcc32 as is? I think it'd be a problem if we change it to jcc8 and the condition is not met. As we relocate jump with a branch table in the trampoline, the upper bytes of offset should be zero. So we can change LSB even in jcc32 in `mcount_arch_patch_branch()` IMO.

Can you please give more details about the problem when the condition is not met.
I think if the condition is not met, the next instruction will be executed. Since handle_rel_jmp() returns the size of the jcc8 insn (2 bytes in this case), the remaining bytes of the buffer won't be used.

Reply:
Hmm.. you changed jcc32 (5 bytes) to jcc8 (2 bytes). So when the condition is not met, the instruction should move to +5 instead of +2, right?

Reply:
Maybe it doesn't matter since it won't add more instruction after the jcc32..

Reply:
Nevermind, I was confused that we don't change it in-place..

Reply:
Speaking about changing jcc32 and jcc8; to be sure that our jcc will always reach the trampoline, we can always use jcc32. If it's a jcc32 we keep it. else (If it's a jcc8) we upgrade it to jcc32. Do you think it's worth it ?

Reply:
I think jcc8 is fine.